tlvc/src/lib.rs

```diff
@@ -363,22 +408,36 @@ impl<R> ChunkHandle<R> {
     where
         R: TlvcRead,
     {
+        // Caclulate the body checksum.
         let mut c = begin_body_crc();
-        let end = self.body_position + self.header.len.get() as u64;
-        let mut pos = self.body_position;
-        while pos != end {
-            let portion = usize::try_from(end - pos)
-                .unwrap_or(usize::MAX)
-                .min(buffer.len());
-            self.source.read_exact(pos, &mut buffer[..portion])?;
-            c.update(&buffer[..portion]);
-            pos += u64::try_from(portion).unwrap();
+        let mut pos = usize::try_from(self.body_position)
+            .ok()
+            .ok_or(TlvcReadError::Truncated)?;
+        let contents_len = usize::try_from(self.header.len.get())
+            .ok()
+            .ok_or(TlvcReadError::Truncated)?;
+        let end = pos
+            .checked_add(contents_len)
+            .ok_or(TlvcReadError::Truncated)?;
+        let len = buffer.len();
+        while pos < end {
+            let portion = (end - pos).min(len);
+            let buf = &mut buffer[..portion];
+            self.source.read_exact(
+                u64::try_from(pos).ok().ok_or(TlvcReadError::Truncated)?,
```

Comment:
Again, this could be 
```suggestion
                u64::try_from(pos).map_err(|_| TlvcReadError::Truncated)?,
```
unless there's a reason to write it this way I wasn't aware of?